as.c

```diff
@@ -32,11 +33,338 @@
 
 #define URL_LEN 1024
 #define PROFILE_NAME_LEN 128
+#define READ_CHUNK_LEN 65536
 
+#define PROFILE_NAME_TOKEN L"# OVPN_ACCESS_SERVER_PROFILE="
+#define FRIENDLY_NAME_TOKEN L"# OVPN_ACCESS_SERVER_FRIENDLY_NAME="
+
+/**
+ * Extract profile name from profile content.
+ *
+ * Profile name is either (sorted in priority order):
+ * - value of OVPN_ACCESS_SERVER_FRIENDLY_NAME
+ * - value of OVPN_ACCESS_SERVER_PROFILE
+ * - URL
+ *
+ * @param profile profile content
+ * @param url profile URL, might be used for name
+ * @param out_name extracted profile name
+ * @param out_name_length max length of out_name char array
+ */
+void
+ExtractProfileName(const WCHAR *profile, const WCHAR *url, WCHAR *out_name, size_t out_name_length)
+{
+    WCHAR friendly_name[PROFILE_NAME_LEN] = { 0 };
+    WCHAR profile_name[PROFILE_NAME_LEN] = { 0 };
+
+    /* wcstok() modifies string, need to make a copy */
+    WCHAR *buf = _wcsdup(profile);
+
+    WCHAR *pch = NULL;
+    pch = wcstok(buf, L"\r\n");
+
+    while (pch != NULL) {
+        if (wcsbegins(pch, PROFILE_NAME_TOKEN)) {
+            wcsncpy(profile_name, pch + wcslen(PROFILE_NAME_TOKEN), PROFILE_NAME_LEN - 1);
+            profile_name[PROFILE_NAME_LEN - 1] = L'\0';
+        }
+        else if (wcsbegins(pch, FRIENDLY_NAME_TOKEN)) {
+            wcsncpy(friendly_name, pch + wcslen(FRIENDLY_NAME_TOKEN), PROFILE_NAME_LEN - 1);
+            friendly_name[PROFILE_NAME_LEN - 1] = L'\0';
+        }
+
+        pch = wcstok(NULL, L"\r\n");
+    }
+
+    if (wcslen(friendly_name) > 0)
+        swprintf(out_name, out_name_length, L"%ls.ovpn", friendly_name);
+    else if (wcslen(profile_name) > 0)
+        swprintf(out_name, out_name_length, L"%ls.ovpn", profile_name);
+    else
+        swprintf(out_name, out_name_length, L"%ls.ovpn", url);
+
+    out_name[out_name_length - 1] = L'\0';
+
+    /* sanitize profile name */
+    while (*out_name) {
+        wchar_t c = *out_name;
+        if (c == L'<' || c == L'>' || c == L':' || c == L'\"' || c == L'/' || c == L'\\' || c == L'|' || c == L'?' || c == L'*')
+            *out_name = L'_';
+        ++out_name;
+    }
+
+    free(buf);
+}
+
+void
+ShowWinInetError(HANDLE hWnd)
+{
+    WCHAR err[256] = { 0 };
+    FormatMessageW(FORMAT_MESSAGE_FROM_HMODULE | FORMAT_MESSAGE_FROM_SYSTEM, GetModuleHandleW(L"wininet.dll"),
+        GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), err, _countof(err), NULL);
+    ShowLocalizedMsgEx(MB_OK, hWnd, _T(PACKAGE_NAME), IDS_ERR_URL_IMPORT_PROFILE, GetLastError(), err);
+}
+
+struct UrlComponents
+{
+    int port;
+    WCHAR host[URL_LEN];
+    bool https;
+};
+
+/**
+* Extracts protocol, port and hostname from URL
+*
+* @param url URL to parse, length must be less than URL_MAX
+*/
+void
+ParseUrl(const WCHAR *url, struct UrlComponents* comps)
+{
+    ZeroMemory(comps, sizeof(struct UrlComponents));
+
+    int domain_off = 0;
```

Comment:
Done. Although this might confuse a user which tries to connect via http and gets an error - we don't indicate that we force https. Maybe add uneditable editbox (sounds good, no?) with "https://" just at the left side of editbox with url. OTOH that might look ugly.

Reply:
It'll behave like sites that automatically redirect to https. Looks fine to me as is. 

We could prefill with https:// and make that part not editable. There are other possible improvements including stripping leading whitespace, but let's get this merged and see how it fares in real use.